### PR TITLE
Pup up

### PR DESCRIPTION
**Pull request by stillNiceTanya**

### CODE REVIEW

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 1]
```diff
@@ -1,19 +1,82 @@
...
+      <div className="popup-preloader-wrapper">
+        {showPopup && <LocationPopUp options={areasData} />}
+        {showPopup && !(areasData.length > 0) && (
+          <img
+            className="preloader-img"
+            src={preloader}
+            alt=""
+          />
+        )}
+      </div>
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:78]: You can try this
```suggestion
        {showPopup && (areasData.length > 0 
            ? <LocationPopUp options={areasData} /> 
            : <img
                className="preloader-img"
                src={preloader}
                alt=""
              />
          )}
      </div>
```

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 2]
```diff
@@ -1,19 +1,82 @@
...
+
+  const handleTogglePopup = () => {
+    setShowPopup(!showPopup);
+  };
+
+  const handlePopupOutsideClick = () => {
+    if (!showPopup) {
+      return;
+    }
+    setShowPopup(!showPopup);
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:23]:
```suggestion
   setShowPopup(false)
```

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.css [hunk 3]
```diff
@@ -1,9 +1,24 @@
...
   margin-right: 5px;
 }
+
+.preloader-img {
+  z-index: 999;
+  position: absolute;
+  top: 185px;
+  left: 125px;
+  width: 50px;
+  height: 50px;
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.css:19]: Looks like you can do this with flex, without absolute position and left and top params

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// при сохранении(только когда что-то выбрано, если не выбрано, то кнопка сохранить нерабочая), список сохраненных городов появляется в списке компонента lacatino-select
+// при сохранении положить список выбранных город в локал сторадж
+// кастомизировать попап красивым :)
+
+export default function LocationPopUp({ options }) {
+  // const options = [
+  //   { value: "chocolate", label: "Chocolate" },
+  //   { value: "strawberry", label: "Strawberry" },
+  //   { value: "vanilla", label: "Vanilla" },
+  // ];
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:17]: Do you still need these comments

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx (hunk 1 above)
Comment (stillNiceTanya) [reply src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:78]: in that case we can see  preloader before <LocationPopUp> appear. But  preloader should appear on <LocationPopUp> area

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx (hunk 4 above)
Comment (stillNiceTanya) [reply src/components/Header/components/LocationPopUp/LocationPopUp.jsx:17]: yep, I haven't asana, I need it

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
-import Select from "react-select";
-
+import Select, { components } from "react-select";
+// import Tooltip from "@atlaskit/tooltip";
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:2]: commented unused code
please, remove

File: src/components/Header/components/LocationPopUp/LocationPopUp.css [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+.pop-up-button.active {
+  background-color: #2684ff;
+}
+
+.pop-up-button.passive {
+  background-color: rgb(226, 223, 223);
+  cursor: not-allowed;
+}
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.css:34]: Looks like unused class

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  onSave,
+}) {
+  const multiValueContainer = ({ selectProps, data }) => {
+    const label = data.label;
+    const allSelected = selectProps.value;
+    const index = allSelected.findIndex((selected) => selected.label === label);
+    const isFirstSelected = index === 0;
+    const labelSuffix = isFirstSelected ? ` (${allSelected.length})` : ", ";
+    const val = `${labelSuffix}${label}`;
+    return val;
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:26]:
```suggestion
    const labelPrefix = isFirstSelected ? `(${allSelected.length}) ` : ", ";
    const val = `${labelPrefix}${label}`;
    
    return val;
```

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  };
+
+  const ControlComponent = (props) => (
+    <div style={controlStyles}>
+      <components.Control {...props} />
+    </div>
+  );
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:33]: Let's declare this component outside `LocationPopUp` component next to `controlStyles`

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  borderRadius: "10px 10px 0 0",
+};
+
+export default function LocationPopUp({
+  options,
+  selectedCities,
+  onChange,
+  onSave,
+}) {
+  const multiValueContainer = ({ selectProps, data }) => {
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:19]: It is better to name React components in Capitalized style
```suggestion
  const MultiValueContainer = ({ selectProps, data }) => {
```

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        closeMenuOnSelect={false}
+        isMulti
+        defaultValue={selectedCities}
+        onChange={onChange}
+        components={{
+          MultiValueContainer: multiValueContainer,
+          Control: ControlComponent,
+          IndicatorsContainer: () => null,
+        }}
+        IndicatorsContainer={{ isDisabled: true }}
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:50]: Is it the correct prop? It looks like it capitalized

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        menuIsOpen={true}
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:51]:
```suggestion
        menuIsOpen
```

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            alignItems: "flex-start",
+          }),
+
+          menu: (base) => ({
+            ...base,
+            marginTop: 0,
+            marginBottom: 0,
+            paddingTop: 0,
+            borderRadius: 0,
+          }),
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:89]: **praise:** 
```suggestion
          }),
          valueContainer: (base) => ({
            ...base,
            textOverflow: "ellipsis",
            maxWidth: "90%",
            whiteSpace: "nowrap",
            overflow: "hidden",
            display: "flex",
            flexDirection: "column",
            alignItems: "flex-start",
          }),
          menu: (base) => ({
            ...base,
            marginTop: 0,
            marginBottom: 0,
            paddingTop: 0,
            borderRadius: 0,
          }),
```

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        menuIsOpen={true}
+        downChevron={false}
+        autoFocus
+        placeholder="Регион или город"
+        maxMenuHeight="300px"
+        styles={{
+          control: (baseStyles, state) => ({
+            ...baseStyles,
+            borderRadius: "15px",
+            borderColor: state.isFocused ? "grey" : "grey",
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:60]: Do we need a ternary operator here? It looks like we use the same color

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+            marginBottom: 0,
+            paddingTop: 0,
+            borderRadius: 0,
+          }),
+        }}
+      />
+
+      <Button
+        onClick={onSave}
+        className="pop-up-button active"
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:95]: It looks like we don't need `active` class anymore

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 15]
```diff
@@ -1,19 +1,115 @@
...
 import { faCompass } from "@fortawesome/free-regular-svg-icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { useState, useEffect } from "react";
+import LocationPopUp from "../../LocationPopUp/LocationPopUp";
+import preloader from "./img/loading-loader.svg";
+import OutsideClick from "../../hooks/useOutsideClick";
 
 import "./LocationSelectButton.css";
 
+const SELECT_VALUE_KEY = "MySelectValue";
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:10]:
```suggestion
const selectedLocationLocalStorageKey = "MySelectValue";
```

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 16]
```diff
@@ -1,19 +1,115 @@
...
+
+      let areas = json.reduce((all, el) => {
+        all.push({ value: el.id, label: el.name });
+        if (el.cities) {
+          all.push(
+            ...el.cities.map((item) => ({ value: item.id, label: item.name }))
+          );
+        }
+        return all;
+      }, []);
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:71]: It is better to create a separate function to get and transform data from the API

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 17]
```diff
@@ -1,19 +1,115 @@
...
+
+  const handlePopupOutsideClick = () => {
+    setShowPopup(false);
+  };
+
+  useEffect(() => {
+    const requestOptions = {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+    };
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:53]: It is better to use these options in a separate function for getting data from API

File: src/components/Header/components/LocationPopUp/LocationPopUp.css [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+  width: 300px;
+  height: 40px;
+  background-color: #2684ff;
+  border: none;
+  bottom: 0;
+  box-shadow: 0 0 0 1px hsl(0deg 0% 0% / 10%), 0 4px 11px hsl(0deg 0% 0% / 10%);
+  border-radius: 0 0 4px 4px;
+}
+
+.pop-up-button.active:active {
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.css:22]: Looks like we don't use `.active` class anymore, let's check it and get rid of it if so

File: src/components/Header/components/LocationPopUp/LocationPopUp.jsx [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+}) {
+  const MultiValueContainer = ({ selectProps, data }) => {
+    const label = data.label;
+    const allSelected = selectProps.value;
+    const index = allSelected.findIndex((selected) => selected.label === label);
+    const isFirstSelected = index === 0;
+    const labelPrefix = isFirstSelected ? ` (${allSelected.length})` : ", ";
+    const val = `${labelPrefix}${label}`;
+    return val;
+  };
```
Comment (Sponom) [inline src/components/Header/components/LocationPopUp/LocationPopUp.jsx:33]: We can create this component outside of `LocationPopUp` component as we did with `ControlComponent`

File: src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx [hunk 20]
```diff
@@ -1,19 +1,104 @@
...
+    if (!showPopup || areasData.length > 0) {
+      return;
+    }
+
+    fetchAreas()
+      .then((data) => {
+        setData(data);
+      })
+      .catch(console.error);
+  }, [showPopup, areasData]); //react
```
Comment (Sponom) [inline src/components/Header/components/TopBar/LocationSelectButton/LocationSelectButton.jsx:65]: Unused comment

**Pull request merged**